File: sdk/aipp-python/aipp/middleware.py
import hashlib
import time
import json
import hmac
from fastapi import Request, Response
from starlette.middleware.base import BaseHTTPMiddleware
from .client import Aipp
from .jwt import sign_jwt, verify_jwt

class L402Middleware(BaseHTTPMiddleware):
    def __init__(
        self,
        app,
        client: Aipp,
        jwt_secret: str,
        resource_id: str,
        amount_sats: int = None,
        amount_usd: float = None,
        expires_in_seconds: int = 3600
    ):
        super().__init__(app)
        self.client = client
        self.jwt_secret = jwt_secret
        self.resource_id = resource_id
        self.amount_sats = amount_sats
        self.amount_usd = amount_usd
        self.expires_in_seconds = expires_in_seconds
        
        if self.amount_sats is None and self.amount_usd is None:
            raise ValueError("Either amount_sats or amount_usd must be provided")

    async def dispatch(self, request: Request, call_next):
        auth_header = request.headers.get("Authorization")
        valid = False
        
        if auth_header and auth_header.startswith("L402 "):
            parts = auth_header[5:].split(":")
            if len(parts) == 2:
                macaroon_str, preimage = parts
                try:
                    payload = verify_jwt(macaroon_str, self.jwt_secret)
                    if payload.get("resource_id") == self.resource_id:
                        preimage_hash = hashlib.sha256(bytes.fromhex(preimage)).hexdigest()
                        expected_hash = payload.get("payment_hash")
                        if expected_hash and hmac.compare_digest(preimage_hash, expected_hash):
                            valid = True
                except Exception:
                    # Invalid, expired, or tampered JWT. Fall through to 402 to issue a new invoice.
                    pass
        
        if valid:
            return await call_next(request)
            
        try:
            charge = self.client.create_charge(
                amount_sats=self.amount_sats,
                amount_usd=self.amount_usd,
                memo=f"L402 Payment for {self.resource_id}"
            )
            
            payload = {
                "payment_hash": charge.payment_hash,
                "resource_id": self.resource_id,
                "exp": int(time.time()) + self.expires_in_seconds
            }
            jwt_token = sign_jwt(payload, self.jwt_secret)
            
            headers = {
                "Www-Authenticate": f'L402 macaroon="{jwt_token}" invoice="{charge.payment_request}"'
            }
            content = {
                "error": "Payment Required",
                "code": "L402",
                "payment_request": charge.payment_request,
                "macaroon": jwt_token
            }
            return Response(
                content=json.dumps(content), 
                status_code=402, 
                headers=headers, 
                media_type="application/json"
            )
        except Exception as e:
            return Response(
                content=json.dumps({"error": "Failed to generate L402 challenge", "details": str(e)}), 
                status_code=500, 
                media_type="application/json"
            )

class DualMiddleware(BaseHTTPMiddleware):
    def __init__(
        self,
        app,
        client: Aipp,
        jwt_secret: str,
        resource_id: str,
        amount_sats: int = None,
        amount_usd: float = None,
        expires_in_seconds: int = 3600
    ):
        super().__init__(app)
        self.client = client
        self.jwt_secret = jwt_secret
        self.resource_id = resource_id
        self.amount_sats = amount_sats
        self.amount_usd = amount_usd
        self.expires_in_seconds = expires_in_seconds
        
        if self.amount_sats is None and self.amount_usd is None:
            raise ValueError("Either amount_sats or amount_usd must be provided")

    async def dispatch(self, request: Request, call_next):
        auth_header = request.headers.get("Authorization")
        tx_hash = request.query_params.get("tx_hash") or request.headers.get("payment-signature") or request.headers.get("x-payment-signature")
        payment_hash = request.query_params.get("payment_hash") or request.headers.get("x-payment-hash")
        valid = False
        
        # 1. Try L402 Verification
        if auth_header and auth_header.startswith("L402 "):
            parts = auth_header[5:].split(":")
            if len(parts) == 2:
                macaroon_str, preimage = parts
                try:
                    payload = verify_jwt(macaroon_str, self.jwt_secret)
                    if payload.get("resource_id") == self.resource_id:
                        preimage_hash = hashlib.sha256(bytes.fromhex(preimage)).hexdigest()
                        expected_hash = payload.get("payment_hash")
                        if expected_hash and hmac.compare_digest(preimage_hash, expected_hash):
                            valid = True
                except Exception:
                    pass
        
        # 2. Try x402 Verification if not valid
        if not valid:
            if not tx_hash and auth_header:
                if auth_header.startswith("Bearer "):
                    tx_hash = auth_header[7:].strip()
                elif auth_header.startswith("x402 "):
                    tx_hash = auth_header[5:].strip()
            
            if tx_hash:
                if not payment_hash and auth_header and auth_header.startswith("L402 "):
                    parts = auth_header[5:].split(":")
                    try:
                        payload = verify_jwt(parts[0], self.jwt_secret)
                        payment_hash = payload.get("payment_hash")
                    except Exception:
                        pass
                
                if payment_hash and tx_hash:
                    try:
                        charge_status = self.client.get_charge(payment_hash, tx_hash)
                        if charge_status.status == "settled":
                            valid = True
                    except Exception:
                        pass

        if valid:
            return await call_next(request)
            
        # 3. Generate DUAL Challenge
        try:
            charge = self.client.create_charge(
                amount_sats=self.amount_sats,
                amount_usd=self.amount_usd,
                protocol="dual",
                memo=f"Dual-Rail Payment for {self.resource_id}"
            )
            
            payload = {
                "payment_hash": charge.payment_hash,
                "resource_id": self.resource_id,
                "exp": int(time.time()) + self.expires_in_seconds
            }
            jwt_token = sign_jwt(payload, self.jwt_secret)
            
            # Generate x402 challenge object
            price_val = charge.amount_usd or self.amount_usd or 0.01
            challenge_obj = {
                "scheme": "exact",
                "network": charge.network or "base",
                "payTo": charge.pay_to or "",
                "price": f"{price_val:.2f}",
                "token": charge.token or "",
                "payment_hash": charge.payment_hash
            }
            challenge_base64 = hashlib.base64 = hmac.base64 = json.dumps(challenge_obj).encode("utf-8")
            import base64
            challenge_b64str = base64.b64encode(challenge_base64).decode("utf-8")
            
            headers = {
                "Www-Authenticate": f'L402 macaroon="{jwt_token}" invoice="{charge.payment_request or ""}"',
                "PAYMENT-REQUIRED": challenge_b64str
            }
            content = {
                "error": "Payment Required",
                "code": "402",
                "payment_hash": charge.payment_hash,
                "pricing": {
                    "usd": charge.amount_usd or self.amount_usd or None,
                    "sats": charge.amount_sats or self.amount_sats or None
                },
                "payment_methods": {
                    "lightning": {
                        "protocol": "L402",
                        "payment_request": charge.payment_request or None,
                        "macaroon": jwt_token
                    },
                    "usdc_base": {
                        "protocol": "x402",
                        "pay_to": challenge_obj["payTo"],
                        "token": challenge_obj["token"],
                        "network": challenge_obj["network"],
                        "amount_usd": float(challenge_obj["price"])
                    }
                },
                "instructions": "Pay the Lightning invoice and supply the preimage in 'Authorization: L402 macaroon:preimage' OR transfer the USDC amount to 'pay_to' and supply the transaction hash in 'Authorization: Bearer tx_hash' or 'payment-signature' header."
            }
            return Response(
                content=json.dumps(content), 
                status_code=402, 
                headers=headers, 
                media_type="application/json"
            )
        except Exception as e:
            return Response(
                content=json.dumps({"error": "Failed to generate Dual-Rail challenge", "details": str(e)}), 
                status_code=500, 
                media_type="application/json"
            )

# ============================================================================
# Roadmap v2.0 - High-Level Problem-Oriented Middleware ("Set & Forget")
# ============================================================================

import os
import functools

def parse_price(price):
    if isinstance(price, (int, float)):
        return {"amount_usd": float(price)}
    clean = str(price).strip().lower()
    if clean.endswith("sats") or clean.endswith("sat"):
        val = clean.replace("sats", "").replace("sat", "").strip()
        return {"amount_sats": int(val) if val.isdigit() else 100}
    clean_usd = clean.replace("$", "").strip()
    try:
        return {"amount_usd": float(clean_usd)}
    except ValueError:
        return {"amount_usd": 0.01}

def _get_client(custom_client=None):
    if custom_client:
        return custom_client
    api_key = os.getenv("AIPP_KEY") or os.getenv("AIPP_API_KEY")
    if not api_key:
        raise ValueError("AIPP Error: AIPP_KEY environment variable is missing. Set AIPP_KEY in .env or pass client.")
    base_url = os.getenv("AIPP_API_URL") or "https://aipp.dev"
    return Aipp(api_key=api_key, base_url=base_url)

def _get_jwt_secret(custom_secret=None):
    return custom_secret or os.getenv("AIPP_JWT_SECRET") or os.getenv("AIPP_KEY") or "aipp_default_secret_key"

def protect_agent(price=0.01, resource_id=None, client=None, jwt_secret=None):
    """
    Decorator for FastMCP or AI Agent tool endpoints.
    """
    def decorator(func):
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            return func(*args, **kwargs)
        wrapper.__aipp_protected__ = True
        wrapper.__aipp_price__ = price
        wrapper.__aipp_resource_id__ = resource_id or func.__name__
        return wrapper
    return decorator

def protect_api(price=0.01, resource_id=None, client=None, jwt_secret=None):
    """
    Helper to construct DualMiddleware for FastAPI / Starlette applications.
    """
    parsed = parse_price(price)
    c = _get_client(client)
    secret = _get_jwt_secret(jwt_secret)
    res_id = resource_id or "api_endpoint"
    
    def middleware_factory(app):
        return DualMiddleware(
            app=app,
            client=c,
            jwt_secret=secret,
            resource_id=res_id,
            amount_usd=parsed.get("amount_usd"),
            amount_sats=parsed.get("amount_sats")
        )
    return middleware_factory

def protect_content(price=0.10, resource_id=None, client=None, jwt_secret=None):
    return protect_api(price=price, resource_id=resource_id or "content", client=client, jwt_secret=jwt_secret)

def protect_download(price=1.00, resource_id=None, client=None, jwt_secret=None):
    return protect_api(price=price, resource_id=resource_id or "download", client=client, jwt_secret=jwt_secret)

File: sdk/aipp-node/src/middleware.ts
import crypto from 'crypto';
import { Aipp } from './client';
import { signJwt, verifyJwt } from './jwt';

export interface L402Options {
  client: Aipp;
  jwtSecret: string;
  resourceId: string;
  amountSats?: number;
  amountUsd?: number;
  expiresInSeconds?: number;
}

export function l402Paywall(options: L402Options) {
  if (!options.amountSats && !options.amountUsd) {
    throw new Error('Either amountSats or amountUsd must be provided for l402Paywall');
  }

  return async (req: any, res: any, next: any) => {
    const authHeader = req.headers.authorization || req.headers.Authorization;
    let valid = false;

    if (authHeader && typeof authHeader === 'string' && authHeader.startsWith('L402 ')) {
      const parts = authHeader.substring(5).split(':');
      if (parts.length === 2) {
        const [macaroonStr, preimage] = parts;
        try {
          // Verify JWT signature and expiration
          const payload = verifyJwt(macaroonStr, options.jwtSecret);
          
          if (payload.resource_id !== options.resourceId) {
            throw new Error('Invalid resource');
          }

          // Verify preimage hash matches payment_hash securely
          const preimageHash = crypto.createHash('sha256').update(Buffer.from(preimage, 'hex')).digest('hex');
          const expectedHash = payload.payment_hash;
          
          if (
            typeof expectedHash === 'string' &&
            expectedHash.length === preimageHash.length &&
            crypto.timingSafeEqual(Buffer.from(preimageHash, 'hex'), Buffer.from(expectedHash, 'hex'))
          ) {
            valid = true;
          }
        } catch (e) {
          // Invalid or expired JWT -> Fall through to 402 and issue new invoice
        }
      }
    }

    if (valid) {
      return next();
    }

    try {
      // Create invoice via AIPP
      const charge = await options.client.createCharge({
        amountSats: options.amountSats,
        amountUsd: options.amountUsd,
        memo: `L402 Payment for ${options.resourceId}`
      });

      // Issue JWT
      const expiresIn = options.expiresInSeconds || 3600; // default 1 hour
      const payload = {
        payment_hash: charge.payment_hash,
        resource_id: options.resourceId,
        exp: Math.floor(Date.now() / 1000) + expiresIn
      };
      
      const jwtToken = signJwt(payload, options.jwtSecret);

      res.status(402);
      res.setHeader('Www-Authenticate', `L402 macaroon="${jwtToken}" invoice="${charge.payment_request}"`);
      res.json({
        error: "Payment Required",
        code: "L402",
        payment_request: charge.payment_request,
        macaroon: jwtToken
      });
    } catch (err: any) {
      res.status(500).json({ error: "Failed to generate L402 challenge", details: err.message || err });
    }
  };
}

export interface X402Options {
  client: Aipp;
  resourceId: string;
  amountUsd: number;
}

export function x402Paywall(options: X402Options) {
  return async (req: any, res: any, next: any) => {
    const authHeader = req.headers.authorization || req.headers.Authorization;
    let txHash = (req.query.tx_hash || req.headers['payment-signature'] || req.headers['x-payment-signature']) as string;
    let paymentHash = (req.query.payment_hash || req.headers['x-payment-hash']) as string;

    if (!txHash && authHeader && typeof authHeader === 'string') {
      if (authHeader.startsWith('Bearer ')) {
        txHash = authHeader.substring(7).trim();
      } else if (authHeader.startsWith('x402 ')) {
        txHash = authHeader.substring(5).trim();
      } else if (authHeader.startsWith('L402 ')) {
        // Fallback or double format check
        const parts = authHeader.substring(5).split(':');
        if (parts.length === 2) {
          paymentHash = parts[0];
          txHash = parts[1];
        }
      }
    }

    if (paymentHash && txHash) {
      try {
        const chargeStatus = await options.client.getCharge(paymentHash, txHash);
        if (chargeStatus.status === 'settled') {
          return next();
        }
      } catch (e) {
        // Validation failed, proceed to challenge
      }
    }

    try {
      const charge = await options.client.createCharge({
        amountUsd: options.amountUsd,
        protocol: 'x402',
        memo: `x402 Payment for ${options.resourceId}`
      });

      const challengeObj = {
        scheme: 'exact',
        network: 'base',
        payTo: charge.pay_to || '',
        price: options.amountUsd.toFixed(2),
        token: charge.token || '',
        payment_hash: charge.payment_hash
      };
      
      const challengeBase64 = Buffer.from(JSON.stringify(challengeObj), 'utf8').toString('base64');
      res.setHeader('PAYMENT-REQUIRED', challengeBase64);

      res.status(402);
      res.json({
        error: "Payment Required",
        code: "x402",
        payment_hash: charge.payment_hash,
        pay_to: challengeObj.payTo,
        price: challengeObj.price,
        token: challengeObj.token,
        network: challengeObj.network
      });
    } catch (err: any) {
      res.status(500).json({ error: "Failed to generate x402 challenge", details: err.message || err });
    }
  };
}

export interface DualOptions {
  client: Aipp;
  jwtSecret: string;
  resourceId: string;
  amountSats?: number;
  amountUsd?: number;
  expiresInSeconds?: number;
}

export function dualPaywall(options: DualOptions) {
  if (!options.amountSats && !options.amountUsd) {
    throw new Error('Either amountSats or amountUsd must be provided for dualPaywall');
  }

  return async (req: any, res: any, next: any) => {
    const authHeader = req.headers.authorization || req.headers.Authorization;
    let txHash = (req.query.tx_hash || req.headers['payment-signature'] || req.headers['x-payment-signature']) as string;
    let paymentHash = (req.query.payment_hash || req.headers['x-payment-hash']) as string;
    let valid = false;

    // 1. Try to verify L402 Header first
    if (authHeader && typeof authHeader === 'string' && authHeader.startsWith('L402 ')) {
      const parts = authHeader.substring(5).split(':');
      if (parts.length === 2) {
        const [macaroonStr, preimage] = parts;
        try {
          const payload = verifyJwt(macaroonStr, options.jwtSecret);
          if (payload.resource_id === options.resourceId) {
            const preimageHash = crypto.createHash('sha256').update(Buffer.from(preimage, 'hex')).digest('hex');
            const expectedHash = payload.payment_hash;
            
            if (
              typeof expectedHash === 'string' &&
              expectedHash.length === preimageHash.length &&
              crypto.timingSafeEqual(Buffer.from(preimageHash, 'hex'), Buffer.from(expectedHash, 'hex'))
            ) {
              valid = true;
            }
          }
        } catch (e) {
          // Fall through to check x402 or re-issue
        }
      }
    }

    // 2. Try to verify Bearer/x402 or query tx_hash next
    if (!valid) {
      if (!txHash && authHeader && typeof authHeader === 'string') {
        if (authHeader.startsWith('Bearer ')) {
          txHash = authHeader.substring(7).trim();
        } else if (authHeader.startsWith('x402 ')) {
          txHash = authHeader.substring(5).trim();
        }
      }

      if (txHash) {
        // If they sent tx_hash, we need the payment_hash to check status.
        // We can extract payment_hash from the L402 header's macaroon if present,
        // or check if it was passed via query/header.
        if (!paymentHash && authHeader && typeof authHeader === 'string' && authHeader.startsWith('L402 ')) {
          const parts = authHeader.substring(5).split(':');
          try {
            const payload = verifyJwt(parts[0], options.jwtSecret);
            paymentHash = payload.payment_hash;
          } catch (e) {}
        }

        if (paymentHash && txHash) {
          try {
            const chargeStatus = await options.client.getCharge(paymentHash, txHash);
            if (chargeStatus.status === 'settled') {
              valid = true;
            }
          } catch (e) {
            // Ignore verification error, will generate new challenge below
          }
        }
      }
    }

    if (valid) {
      return next();
    }

    // 3. Not paid or invalid -> Issue new DUAL challenge
    try {
      const charge = await options.client.createCharge({
        amountSats: options.amountSats,
        amountUsd: options.amountUsd,
        protocol: 'dual',
        memo: `Dual-Rail Payment for ${options.resourceId}`
      });

      // Generate L402 Macaroon
      const expiresIn = options.expiresInSeconds || 3600;
      const payload = {
        payment_hash: charge.payment_hash,
        resource_id: options.resourceId,
        exp: Math.floor(Date.now() / 1000) + expiresIn
      };
      const jwtToken = signJwt(payload, options.jwtSecret);

      // Generate x402 challenge
      const challengeObj = {
        scheme: 'exact',
        network: charge.network || 'base',
        payTo: charge.pay_to || '',
        price: (charge.amount_usd || options.amountUsd || 0.01).toFixed(2),
        token: charge.token || '',
        payment_hash: charge.payment_hash
      };
      const challengeBase64 = Buffer.from(JSON.stringify(challengeObj), 'utf8').toString('base64');

      res.setHeader('Www-Authenticate', `L402 macaroon="${jwtToken}" invoice="${charge.payment_request || ''}"`);
      res.setHeader('PAYMENT-REQUIRED', challengeBase64);
      res.status(402);
      
      res.json({
        error: "Payment Required",
        code: "402",
        payment_hash: charge.payment_hash,
        pricing: {
          usd: charge.amount_usd || options.amountUsd || null,
          sats: charge.amount_sats || options.amountSats || null
        },
        payment_methods: {
          lightning: {
            protocol: "L402",
            payment_request: charge.payment_request || null,
            macaroon: jwtToken
          },
          usdc_base: {
            protocol: "x402",
            pay_to: challengeObj.payTo,
            token: challengeObj.token,
            network: challengeObj.network,
            amount_usd: parseFloat(challengeObj.price)
          }
        },
        instructions: "Pay the Lightning invoice and supply the preimage in 'Authorization: L402 macaroon:preimage' OR transfer the USDC amount to 'pay_to' and supply the transaction hash in 'Authorization: Bearer tx_hash' or 'payment-signature' header."
      });
    } catch (err: any) {
      res.status(500).json({ error: "Failed to generate Dual-Rail challenge", details: err.message || err });
    }
  };
}

// ============================================================================
// Roadmap v2.0 - High-Level Problem-Oriented Middleware ("Set & Forget")
// ============================================================================

export interface BaseProtectOptions {
  /** Price in USD (e.g. 0.01 or "$0.01") or Sats (e.g. "100 sats") */
  price: number | string;
  /** Unique ID for the resource. Defaults to request URL. */
  resourceId?: string;
  /** Optional custom Aipp client instance. Defaults to process.env.AIPP_KEY */
  client?: Aipp;
  /** Optional JWT secret for L402 token signing. Defaults to process.env.AIPP_JWT_SECRET or AIPP_KEY */
  jwtSecret?: string;
  /** Expiration time in seconds for the paid token (Default: 3600s / 1h) */
  expiresInSeconds?: number;
}

export function parsePrice(price: number | string): { amountUsd?: number; amountSats?: number } {
  if (typeof price === 'number') {
    return { amountUsd: price };
  }
  const clean = price.trim().toLowerCase();
  if (clean.endsWith('sats') || clean.endsWith('sat')) {
    const sats = parseInt(clean.replace(/sats?/, '').trim(), 10);
    return { amountSats: isNaN(sats) ? 100 : sats };
  }
  const num = parseFloat(clean.replace('$', '').trim());
  return { amountUsd: isNaN(num) ? 0.01 : num };
}

function getOrCreateClient(customClient?: Aipp): Aipp {
  if (customClient) return customClient;
  const apiKey = process.env.AIPP_KEY || process.env.AIPP_API_KEY;
  if (!apiKey) {
    throw new Error('AIPP Error: AIPP_KEY environment variable is missing. Set AIPP_KEY in .env or pass client instance to protect function.');
  }
  const baseUrl = process.env.AIPP_API_URL || 'https://aipp.dev';
  return new Aipp({ apiKey, baseUrl });
}

function getJwtSecret(customSecret?: string): string {
  return customSecret || process.env.AIPP_JWT_SECRET || process.env.AIPP_KEY || 'aipp_default_secret_key_change_in_prod';
}

/**
 * Protects an API route with micro-payments.
 * Example: app.post('/v1/ai', protectApi({ price: 0.01 }), handler);
 */
export function protectApi(options: BaseProtectOptions) {
  const { amountUsd, amountSats } = parsePrice(options.price);
  return (req: any, res: any, next: any) => {
    const client = getOrCreateClient(options.client);
    const jwtSecret = getJwtSecret(options.jwtSecret);
    const resourceId = options.resourceId || req.originalUrl || req.url || 'api_endpoint';

    return dualPaywall({
      client,
      jwtSecret,
      resourceId,
      amountUsd,
      amountSats,
      expiresInSeconds: options.expiresInSeconds || 3600
    })(req, res, next);
  };
}

/**
 * Protects an AI Agent / MCP Server tool endpoint.
 * Optimized for machine-to-machine HTTP 402 challenges.
 */
export function protectAgent(options: BaseProtectOptions) {
  const { amountUsd, amountSats } = parsePrice(options.price);
  return (req: any, res: any, next: any) => {
    const client = getOrCreateClient(options.client);
    const jwtSecret = getJwtSecret(options.jwtSecret);
    const resourceId = options.resourceId || req.originalUrl || req.url || 'mcp_tool';

    return dualPaywall({
      client,
      jwtSecret,
      resourceId,
      amountUsd,
      amountSats,
      expiresInSeconds: options.expiresInSeconds || 86400
    })(req, res, next);
  };
}

/**
 * Protects premium web content / HTML paywalls.
 */
export function protectContent(options: BaseProtectOptions) {
  const { amountUsd, amountSats } = parsePrice(options.price);
  return (req: any, res: any, next: any) => {
    const client = getOrCreateClient(options.client);
    const jwtSecret = getJwtSecret(options.jwtSecret);
    const resourceId = options.resourceId || req.originalUrl || req.url || 'content_page';

    return dualPaywall({
      client,
      jwtSecret,
      resourceId,
      amountUsd,
      amountSats,
      expiresInSeconds: options.expiresInSeconds || 86400
    })(req, res, next);
  };
}

/**
 * Protects file download endpoints.
 */
export function protectDownload(options: BaseProtectOptions) {
  const { amountUsd, amountSats } = parsePrice(options.price);
  return (req: any, res: any, next: any) => {
    const client = getOrCreateClient(options.client);
    const jwtSecret = getJwtSecret(options.jwtSecret);
    const resourceId = options.resourceId || req.originalUrl || req.url || 'download_asset';

    return dualPaywall({
      client,
      jwtSecret,
      resourceId,
      amountUsd,
      amountSats,
      expiresInSeconds: options.expiresInSeconds || 3600
    })(req, res, next);
  };
}

